Implement transitive dependency resolution using Aether

petroseskinder opened the pull request:
In particular, given a set of artifacts, this CL produces a set of Bazel maven_jar rules. Transitive dependency resolution is achieved using aether. I rolled my own solution for serializing aether's dependency graph into a set of build rules. This is accompanied with a set of unit tests

This has not been integrated into the generate_workspace class. It's a slightly better use of time to first figure out how to parse maven artifacts from a set of pom files. 

The obscenely large size of the CL is due to changes to `generate_workspace.bzl`

Commit:
Update bzl file to include external dependencies on various aether packages

bazel-io commented:
Can one of the admins verify this patch?

kchodorow commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/ArtifactResolver.java:
```diff
@@ -0,0 +1,83 @@
+  private final static Logger logger = Logger.getLogger(
+      MethodHandles.lookup().lookupClass().getName());
+
+  //TODO(petros): add support for managed dependencies, exclusions, and aliases.
```
Will this be handled "automatically" by Aether's machinery?

petroseskinder replied:
For the most part. Though, there is some massaging one needs to do with user input.

kchodorow commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/ArtifactResolver.java:
```diff
@@ -0,0 +1,83 @@
+    try {
+      dependencyResult = system.resolveDependencies(systemSession, dependencyRequest);
+    } catch (DependencyResolutionException e) {
+      //FIXME(petros): This is very fragile. If even one artifact does not resolve, then no artifacts resolve.
```
You could remove the "return null".

petroseskinder replied:
Little change. I return `dependencyResult.getRoot()` and `dependencyResult` is assigned within the try block. I would need to at some point, return null.

kchodorow replied:
It looks like you can get the dependency result from the exception, so you can still return it (and show which dependencies did not resolve).

Please test.

kchodorow commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/ArtifactResolver.java:
```diff
@@ -0,0 +1,83 @@
+      logger.warning("Unable to resolve transitive dependencies: " + e.getMessage());
+      return null;
+    }
+    return dependencyResult.getRoot();
```
What is the root when multiple dependencies are requested?

petroseskinder replied:
The root is an empty sentinel node. Its children are the requested dependencies. This way dependencies are not resolved independently.

The key line is 

`CollectRequest collectRequest = new CollectRequest(roots, managedDependencies, remoteRepositories)`

Added a comment for clarity.

kchodorow commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/ArtifactResolver.java:
```diff
@@ -0,0 +1,83 @@
+   * i.e. a request all transitive dependencies. This request in particular,
+   * ignores any dependencies which are not scoped as Compile
+   */
+  private DependencyRequest createDependencyRequest(List<String> artifacts) {
```
Please rename artifacts -> "coordinates"

kchodorow commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/MavenJarRule.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  public String getRepository() {
+    System.out.println("REPOSITORY " + node.getRepositories());
```
Debugging

kchodorow commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/Rule.java:
```diff
@@ -16,18 +16,18 @@
 /**
  * A struct representing the fields of maven_jar to be written to the WORKSPACE file.
  */
+//TODO(petros): Kill this after refactoring resolvers.
```
Why did you create a new class?

petroseskinder replied:
It does effectively the same thing, except using aether classes. The original class uses maven versions of classes like Artifact and what not. I found that augmenting `Rule` class required a considerable amount of refactoring across the codebase. 

Since I am going to rip the resolvers out, I thought it would be a better use of time to keep both classes.

petroseskinder replied:
This is the direction I am moving with [the next pull request](https://github.com/petroseskinder/migration-tooling/commit/14880cc251a907af40c8a696f4ac15ed7a4368d3#diff-376ae4a942b0ed1920d7f8b28cf11751), as far as ripping out the Resolver

kchodorow reviewed (commented):
Generally looks good!

petroseskinder committed:
Automatically create output directory if it doesn't exist #39 (#42)
Imcludes regression tests for when directory can and cannot be
created.

Commit:
Implement dependency resolution and graph serialization using Aether

Includes unit tests for Aether portion. Note, this commit does not
replace the existing the generate_workspace tool's existing resolution
scheme

Commit:
Add statement in README acknowledging aether

petroseskinder commented:
I'm not sure why github is stating there are conflicts within generate_workspace.bzl
As of commit 9880260, there are no conflicts in generate_workspace.bzl.  😞

kchodorow commented:
Checking out this branch, I get:

```
$ git fetch git://github.com/petroseskinder/migration-tooling.git version-fix
remote: Counting objects: 75, done.
remote: Total 75 (delta 16), reused 16 (delta 16), pack-reused 58
Unpacking objects: 100% (75/75), done.
From git://github.com/petroseskinder/migration-tooling
 * branch            version-fix -> FETCH_HEAD
03:17:13 PM ~/gitroot/migration-tooling (master)$ git reset --hard FETCH_HEAD
HEAD is now at bc0918e Add statement in README acknowledging aether
03:17:22 PM ~/gitroot/migration-tooling (master)$ bazel run //generate_workspace -- -a org.apache.maven:maven-artifact:3.5.0
ERROR: /home/kchodorow/gitroot/migration-tooling/generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/BUILD:1:1: no such package '@org_eclipse_aether_aether_util//jar': The repository could not be resolved and referenced by '//generate_workspace/src/main/java/com/google/devtools/build/workspace/maven:maven'.
ERROR: Analysis of target '//generate_workspace:generate_workspace' failed; build aborted.
INFO: Elapsed time: 1.802s
ERROR: Build failed. Not running target.
```

petroseskinder commented:
Ehh. I see. My apologies. This is from merging the generate_workspace.bzl. It should be fine now. I squashed a number of commits and regenerated the file. 

After the repository rule is merged, it makes sense to add this to the gitignore.

kchodorow commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/GraphSerializer.java:
```diff
@@ -0,0 +1,82 @@
+      return false;
+    }
+    boolean isFirstVisit = visited.add(dependencyNode);
+    if (!isFirstVisit) {
```
In my experience, Maven/Aether does not define equals() methods, so probably `new DependencyNode() != new DependencyNode()`.

petroseskinder replied:
Disappointing they never override the equals() method. 

This is a potential performance issue, since we don't want to traverse/accept a node multiple times. It doesn't affect correctness.

I suspect that object.equals() may suffice for identifying equal nodes. 

This suspicion derives from Aether's [TreeDependencyVisitor](https://github.com/sonatype/sonatype-aether/blob/master/aether-util/src/main/java/org/sonatype/aether/util/graph/TreeDependencyVisitor.java) class. 

```java
public boolean visitEnter( DependencyNode node )
    {
        boolean visited = visitedNodes.put( node, Boolean.TRUE ) != null;

        visits.push( Boolean.valueOf( visited ) );

        if ( visited )
        {
            return false;
        }

        return visitor.visitEnter( node );
    }

```
[Related tests](https://github.com/sonatype/sonatype-aether/blob/master/aether-util/src/test/java/org/sonatype/aether/util/graph/TreeDependencyVisitorTest.java)

I will include a TODO, and investigate this further.

kchodorow replied:
As I suspected, this fails:

```
    DependencyNode n1 = dependencyNode("x:y:1");
    DependencyNode n2 = dependencyNode("x:y:1");
    assertThat(n1).isEqualTo(n2);
```

Can you just make `visited` a set of MavenJarRules? That avoids this whole issue.

petroseskinder replied:
Changed.

kchodorow commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/GraphSerializer.java:
```diff
@@ -0,0 +1,82 @@
+  /* Adds parent and child pointers to nodes for later use by serializer. */
+  public boolean visitLeave(DependencyNode dependencyNode) {
+    MavenJarRule rule = getRule(dependencyNode);
+    ruleGraph.successors(rule).forEach(r -> r.addParent(rule));
```
I don't think we want the parents.  I was parsing them for pom.xml files because I needed the variables they defined, I think.  But we don't actually want their dependencies.

petroseskinder replied:
Don't we want to include in the `generate_workspace.bzl` who requested a given artifact?
```python
# something:something.id requested this
native.maven_jar(
)
```
It is much easier doing this bookkeeping here, rather than having to compute this information later.

kchodorow replied:
Oh right, forgot about that.

kchodorow commented on generate_workspace/src/test/java/com/google/devtools/build/workspace/maven/GraphSerializerTest.java:
```diff
@@ -0,0 +1,100 @@
+    MavenJarRule ruleC = new MavenJarRule(nodeC);
+    addDependency(ruleA, ruleA1);
+
+    assertRuleSetContainsExactly(rules, ruleA, ruleA1, ruleB, ruleC);
```
Add tests for duplicate children:

```
DependencyNode node1 = dependencyNode("a:a:1");
DependencyNode node2 = dependencyNode("a:a:1");
```

petroseskinder replied:
Let me make sure I understand your request. You are asking to test that we never get a rule like
```
A --> {"a:a:1", "a:a:1"}
```
If so, I have added a test within the AetherGraphTraversalTest. That is where this condition is enforced.

kchodorow replied:
No, I mean if we get `x:x:1 -> {y:y:1 -> {a:a:1}, z:z:1 -> {a:a:1}}` we don't end up with two `a:a:1` nodes.

petroseskinder replied:
I see. Test for that can be found [here](https://github.com/bazelbuild/migration-tooling/pull/40/commits/69d01192083dfbb8587c127b8753527e441420ff#diff-681fe7c9eb94648ae236eee7b1e13a5dR70)

kchodorow replied:
What I was looking for was:

```
    DependencyNode sentinel = dependencyNode("dummy:dummy:0");
    DependencyNode nodeA = dependencyNode("a:a:1");
    DependencyNode nodeB = dependencyNode("b:b:1");
    DependencyNode child1 = dependencyNode("child:child:1");
    DependencyNode child2 = dependencyNode("child:child:1");

    sentinel.setChildren(ImmutableList.of(nodeA, nodeB));
    nodeA.setChildren(ImmutableList.of(child1));
    nodeB.setChildren(ImmutableList.of(child2));

    Set<MavenJarRule> rules = GraphSerializer.generateBuildRules(sentinel);

    MavenJarRule ruleA = new MavenJarRule(nodeA);
    MavenJarRule ruleB = new MavenJarRule(nodeB);
    MavenJarRule childRule = new MavenJarRule(child1);
    addDependency(ruleA, childRule);
    addDependency(ruleB, childRule);

    assertRuleSetContainsExactly(rules, ruleA, ruleB, childRule);
```

Which fails.

petroseskinder replied:
Yes, failure is expected. We should not confront a situation like this (otherwise we have much more worrisome issues). 

Currently, to construct this graph, we are using aether's built in dependency resolution mechanism. I think it is safe to assume that aether's graph will not have situations like this. As you have pointed out, DefaultDependencyNode, the node used by aether, does not override its equals method. So the only way to assert equality is by object's `equals()` method. 

If the above situation were to occur, things would be broken/faulty much earlier in the pipeline. Were we to define our own dependency resolution mechanism, we could implement a DependencyNode that actually overrides `equals()`. 

Thoughts?

kchodorow replied:
Ok.

Commit:
Adds test for duplicates in Graph Serialization

petroseskinder commented on generate_workspace/src/test/java/com/google/devtools/build/workspace/maven/GraphSerializerTest.java:
```diff
@@ -0,0 +1,100 @@
+    MavenJarRule ruleC = new MavenJarRule(nodeC);
+    addDependency(ruleA, ruleA1);
+
+    assertRuleSetContainsExactly(rules, ruleA, ruleA1, ruleB, ruleC);
```
Let me make sure I understand your request. You are asking to test that we never get a rule like
```
A --> {"a:a:1", "a:a:1"}
```
If so, I have added a test within the AetherGraphTraversalTest. That is where this condition is enforced.

petroseskinder commented on generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/ArtifactResolver.java:
```diff
@@ -0,0 +1,83 @@
+    try {
+      dependencyResult = system.resolveDependencies(systemSession, dependencyRequest);
+    } catch (DependencyResolutionException e) {
+      //FIXME(petros): This is very fragile. If even one artifact does not resolve, then no artifacts resolve.
```
Noted. Can we separate this out to another pull request? 
This pull request doesn't alter the functionality of generate_workspace. I would like to delay until I remove the existing Resolver.

petroseskinder commented:
> It looks like you can get the dependency result from the exception, so you can still return it (and show which dependencies did not resolve).
>
> Please test.

Noted. Can we separate this out to another pull request? Specifically, I would prefer this be in the same PR that rips out Resolver.java

kchodorow reviewed (commented):
Okay.

Commit:
Use maven jar rule rather than Aether's dependency node for visited set

kchodorow commented:
Can you resolve the generate_workspace.bzl conflicts & I'll merge?

Commit:
Merge branch 'master' of https://github.com/bazelbuild/migration-tooling into version-fix

kchodorow commented:
Jenkins, test this please.

petroseskinder commented:
jenkins says yay! :)
